### ISSUE TITLE
deal.II 8.3 planning

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#1098]
we would like to finish the next deal.II release before the deal.II workshop in August. I thought we might collect references to other issues and PRs here.

List of open PRs:
 - ~~#1133 configuration fixes~~
 - ~~#1117 Slepc fixes~~ (postponed)
 - ~~#1090 periodicity~~
 - #1046 clang warnings
 - ~~#1123 ParameterHandler::MultipleSelection~~
 - ~~#1135~~

**Comment by tjhei** [comment]
@bangerth can you please do step -1? I am happy to take it from there.

**Comment by tjhei** [comment]
Thanks.

Things we should wait for: https://github.com/dealii/dealii/issues/1117

Anything else that really needs to go in?

**Comment by tamiko** [comment]
What about #1090 ?

**Comment by tjhei** [comment]
@kronbichler what is the latest Trilinos version that we know works?

**Comment by kronbichler** [comment]
I've been working with 11.15 and 12.1 (developer versions), so I think that all of 11.12, 11.14 and 12.0 should be fine. I've not tested the issue with trilinos_sparsity_pattern.cc (regarding the non-local entries) yet so I would not include it in the release. My recent test results with Trilinos 11.15 show the following data:
http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=9987
(There is some slight deviation in MueLu but I don't see a problem. The other failing tests are because my PETSc installation does not have MUMPS and some serialization problem, none of which related to deal.II.)

@Rombur Any other

**Comment by tjhei** [comment]
> I've not tested the issue with trilinos_sparsity_pattern.cc (regarding the non-local entries) yet so I would not include it in the release.

Can you please explain what this is about?

> I've been working with 11.15 and 12.1 (developer versions), so I think that all of 11.12, 11.14 and 12.0 should be fine.

does this patch https://github.com/tjhei/dealii/commit/9ef7ea0965f2e991069dcf0829981d98a37a69c1  look okay to you?

**Comment by kronbichler** [comment]
> Can you please explain what this [trilinos_sparsity_pattern.cc] is about?
This is an internal remark that we put to the code at some point because we want to use advanced features of the Epetra_FECrsGraph (to avoid re-allocating memory and make assembly with MPI thread-safe for the non-local part of the matrix that is going to be exported to other processors). I will re-evaluate this after the release to be on the safe side.

>does this patch tjhei@9ef7ea0 look okay to you?
Yes, this looks good to me.

**Comment by Rombur** [comment]
The patch looks fine. I have been using Trilinos 12.0 and 12.1 and it is working fine.

**Comment by tjhei** [comment]
I think after #1133 we should be able to create the release branch. If I don't hear objections, I will go ahead with it later today.

**Comment by tamiko** [comment]
@tjhei I restart simserv04 to test the latest version. I suggest we wait for the run to finish before we branch.

So far, I had no time to
 - reenable icc tests on simserv04
 - update to trilinos 12.0.1 on simserv04

**Comment by tjhei** [comment]
I just created the release branch, please go ahead and start testing.

**Comment by tjhei** [comment]
(and we are at step 9/10)

**Comment by tjhei** [comment]
step 11 is done. Tasks right now:
- write announce-8.3
- write release paper
- 10/ Test the branch thoroughly

**Comment by bangerth** [comment]
I wrote most of the announcement. @tamiko and @luca-heltai: can you please add a sentence about the linear operators?

I made a start with the release paper but am hampered that the file can't be compiled with latex right now because it relies on the ansarticle and anslisting style files. Can someone add those to the directory?

**Comment by tjhei** [comment]
Thanks, @bangerth . I added the ANS files.

**Comment by tjhei** [comment]
Did you go through the changelog to find if there are any other features worth mentioning (beside what you already have in the announcement?)?

**Comment by bangerth** [comment]
@tjhei : Yes. Only the linear operator stuff I asked @tamiko  and @luca-heltai about.

Thanks for the style files. ansarticle.sty required a bunch of 'algorithm' packages that aren't on my system. They don't seem to be necessary, so I just commented out the line.

**Comment by bangerth** [comment]
Now we only need to write the release paper. There isn't likely going to be very much to write about, but I'd appreciate if someone could take a stab.

**Comment by tamiko** [comment]
@bangerth : I have it on my radar. I'll complete the announcement this evening

@tjhei : There seems nothing left to do so far - what about a first release candidate?

**Comment by tjhei** [comment]
I will do an RC tomorrow, but there is some weird stuff happening with Intel 15 on my machine: http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=10165

**Comment by tjhei** [comment]
@tamiko do you want to take a look at the LinearOperator stuff on Intel?

**Comment by tamiko** [comment]
Am 15. Jul 2015, 20:51 schrieb Timo Heister <notifications@github.com>:

> I will do an RC tomorrow, but there is some weird stuff happening with
> Intel 15 on my machine:
> http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=10165

*ugh*

It seems that intel is not quite happy with what everybody else
understands as standard conformant C++11.

I'll have a look at it.

Let's not postpone the release candidate - we can merge a fix also after
rc1 is tagged.

**Comment by tjhei** [comment]
Yeah, no worries. I'll take ec1 tomorrow morning.

**Comment by tjhei** [comment]
Release tagged and all files added: https://github.com/dealii/dealii/releases/tag/v8.3.rc1

**Comment by tamiko** [comment]
@tjhei: So, the majority of the linear_operator test failures vanish with 15.0.3?

**Comment by tjhei** [comment]
Well, we had a ton of blocklinearoperator failures (now gone of course), but there is still  http://cdash.kyomu.43-1.org/testDetails.php?test=16831888&build=10165

**Comment by tamiko** [comment]
@tjhei Yes... I'm currently working on it. This seems to be an sfinae bug in intel-15.

**Comment by tjhei** [comment]
Well, I am trying to understand what the heck is happening here: http://cdash.kyomu.43-1.org/testDetails.php?test=16830069&build=10165 
It looks like fe_values[extractor].divergence (i,q) is just wrong. Sadly I can not run in the debugger so I am stuck. If it stays as is, your test doesn't matter at all. Can we drop support for the intel compiler please? :-)

**Comment by tjhei** [comment]
Oh, and we should try gcc 5.2 clang 3.6.2 (both just released). @tamiko do you have a way to set them up? I am too lazy to compile compilers from source and I don't know about ubuntu binaries for gcc 5.2.

**Comment by tamiko** [comment]
> @tamiko do you have a way to set them up?

Yes. I'll take care of this.

**Comment by tamiko** [comment]
@tjhei: A testsuite run for clang-3.6.2 is under way on my laptop.

**Comment by tamiko** [comment]
@tjhei Results for clang 3.6.2 http://cdash.kyomu.43-1.org/buildSummary.php?buildid=10245

**Comment by tjhei** [comment]
Clang 3.6.2 is looking good. Thanks.

**Comment by tjhei** [comment]
Any issues found so far?

**Comment by tamiko** [comment]
@tjhei not with clang 3.6.2 - I recompile part of my userland at the moment. As soon as this is finished, I'll fire up a test with gcc-5.2.0.

**Comment by tamiko** [comment]
@tjhei - gcc-5.2.0 looks pretty fine, too [1]. Ignore the serialization tests - those require an interface with std::string and fail due to incompatible ABIs.

[1] http://cdash.kyomu.43-1.org/buildSummary.php?buildid=10277

**Comment by kronbichler** [comment]
@tjhei @tamiko As listed in #1164, I also checked the pre-release version of clang-3.7. It looks very good and problems are only due to my aggressive FMA settings.

**Comment by tjhei** [comment]
Great. And intel 15 is finally happy too: http://cdash.kyomu.43-1.org/viewTest.php?onlyfailed&buildid=10268

**Comment by tjhei** [comment]
maybe take over #1176 ?

**Comment by tjhei** [comment]
and #1179...

**Comment by bangerth** [comment]
Andrea Bonito ran into the problem fixed in #1180 (and a follow-up in #1188). It's a pretty self-contained issue. Would it be worth moving this to 8.3?

**Comment by luca-heltai** [comment]
I'm in favour of moving  #1180 to 8.3. I also gave a stab at filling the 8.3 paper with the contributions from my group. Feel free to remove if you think it's too much.

**Comment by tjhei** [comment]
Okay, should I do another RC? Do we have anything else open?

**Comment by tamiko** [comment]
> Okay, should I do another RC? Do we have anything else open?

There can never be enough RCs :-)

**Comment by luca-heltai** [comment]
I'm finishing up two PR, which I'd be happy to see in 8.3. The first is #1195, which is self contained in vector_tools.h, and addresses a bug in a new feature just introduced in 8.3. Shipping that feature out with the bug would not make much sense... 

The second is #1191 which fixes warnings in LinearOperator, and that is also a new feature of 8.3... I'm waiting for a few tests to complete, and I'll commit my last change to that one...

**Comment by tjhei** [comment]
will do rc2 after we decide on #1201

**Comment by tamiko** [comment]
If we do not come up with a fix for <code>mapping_fe_field_real_to_unit_b2_mask</code>, I'd suggest to remove this test from the release branch (@luca-heltai).

**Comment by luca-heltai** [comment]
That test passes for me with #1195, which was just merged here as well.

**Comment by luca-heltai** [comment]
I've just tested ce1b1d2 on my mac, and that test is passing for me... let's see what the tester says about it.

**Comment by tjhei** [comment]
okay, rc2 is created. Thanks for the mac packages, @luca-heltai . Everyone, please help test this a bit!

**Comment by bangerth** [comment]
This seems to work for me.

**Comment by tamiko** [comment]
Can we also take over the testsuite fixes
  88ae1feb4a7153cd59016c5edc8b65b34cdba77e
 #1223 
to the release branch?

**Comment by bangerth** [comment]
My suggestion would be to call it a day with the release. We need to come to a point where we don't make any more changes. Surely, whether the testsuite has a couple more or less well understood failures doesn't make much difference.

**Comment by kronbichler** [comment]
One more thing I found while checking the tutorial programs. When compiling step-32, I get the following warning:
```
[ 50%] Building CXX object CMakeFiles/step-32.dir/step-32.cc.o
In file included from /home/kronbichler/deal/deal.II/include/deal.II/base/work_stream.h:22:0,
                 from /home/kronbichler/deal/examples/step-32/step-32.cc:32:
/home/kronbichler/deal/deal.II/include/deal.II/base/multithread_info.h:47:7: warning: ‘dealii::MultithreadInfo::n_cpus’ is deprecated [-Wdeprecated-declarations]
 class MultithreadInfo
       ^
/home/kronbichler/deal/deal.II/include/deal.II/base/multithread_info.h:64:29: note: declared here
   static const unsigned int n_cpus DEAL_II_DEPRECATED;
```
I get this warning on every tutorial program that uses WorkStream, without using the deprecated static variable `MultithreadInfo::n_cpus` variable. The only solution I see is to remove DEAL_II_DEPRECATED from the variable (but not the associated documentation). Ideas?

**Comment by tjhei** [comment]
@kronbichler  I don't see the warning here (gcc 4.8.4). Is it enough to include multithread_info.h to see the warning? If yes, we need to remove the deprecation from the declaration.

**Comment by bangerth** [comment]
Indeed, `work_stream.h` doesn't use `n_cpus` at all. Are you per chance looking at an older version of step-32.cc?

**Comment by kronbichler** [comment]
I'm looking at the dealii-8.3 branch files, current HEAD. I'm on gcc-5.2 so that could explain the difference (not sure what the C++ rulebook says about deprecation warnings which use compiler-specific settings; it could also be unintended behavior of gcc-5.2). I don't see the warning on clang 3.5 and 3.7.

@tjhei Indeed it is enough to include multithread_info.h. I tested it on an otherwise clean step-3.

**Comment by kronbichler** [comment]
I checked on a rather minimal example:
```
struct Foo
{
  static const unsigned int bar __attribute__((deprecated));
};

int test()
{
  return 0;
}
```
gcc-5.1/5.2 warn about a deprecated variable, whereas gcc-4.6, 4.8, 4.9 and clang 3.5, 3.6, 3.7 do not warn. Should we disable the deprecation warning? I guess the gcc-5.x compiler will be quite popular soon as it will ship in several big Linux distros by October fall.

**Comment by bangerth** [comment]
Let's close the book on 8.3 and simply take this as a reason to keep up a 6 month schedule for releases -- most people won't immediately jump to the newest distro the minute it comes out.

The solution here is simply to make sure that we remove the deprecated variable after the release.

**Comment by kronbichler** [comment]
I'm fine with this procedure. Having closer releases is definitely a good idea.

**Comment by tamiko** [comment]
Agreed.

**Comment by bangerth** [comment]
8.3 is out. Did anyone send around the release announcement to the various mailing lists, na-digest, etc?

**Comment by bangerth** [comment]
(Also, once this and possibly any other steps on the task list are done, we can close this issue.)

**Comment by bangerth** [comment]
I think we're through with it all!
